### ISSUE TITLE
Apparently inefficient incremental builds

### ISSUE THREAD

**Issue by jfehrle** [coq/coq#8315]
It seems to me that incremental builds of Coq do a lot of unnecessary work.

After a successful make, suppose I make a change to printing/proof_diffs.ml, but I don't change proof_diffs.mli or any other file.  Obviously we need to link in the new object code for proof_diffs.ml into executables and (because we choose to) some archive files, but I don't see why we need to recompile any other .ml files.  If the .mli file is unchanged, the module presents the same external interface.  So why would we need to rebuild any other .ml file?

Or is there some quirk in OCaml that requires rebuilding other .ml files?

Guessing here: Perhaps part of it is because the compiled .ml and .mli files go into a single .cmxa archive for each directory, e.g. printing.cmxa?  And then dependencies on printing/foo.mli become dependencies on printing.cmxa, which depends on every source file in the printing directory.  (But when I look at .mlfiles.d, the dependencies appear to be very precise.)

If I'm correct, 90% of the following make steps would go away.

```
$ make coqocaml
rm -f
make --warn-undefined-variable --no-builtin-rules -f Makefile.build coqocaml
make[1]: Entering directory '/cygdrive/c/Users/Jim.DESKTOP-MMPNVGH/Documents/coqSource/coq'
OCAMLDEP  MLFILES MLIFILES
OCAMLOPT  printing/proof_diffs.ml
OCAMLOPT  printing/printer.ml
OCAMLOPT  printing/printmod.ml
OCAMLOPT  printing/prettyp.ml
OCAMLOPT -a -o printing/printing.cmxa
OCAMLOPT  tactics/tactics.ml
OCAMLOPT  tactics/elim.ml
OCAMLOPT  tactics/equality.ml
OCAMLOPT  tactics/contradiction.ml
OCAMLOPT  tactics/inv.ml
OCAMLOPT  tactics/leminv.ml
OCAMLOPT  tactics/hints.ml
OCAMLOPT  tactics/auto.ml
OCAMLOPT  tactics/eauto.ml
OCAMLOPT  tactics/class_tactics.ml
OCAMLOPT  tactics/eqdecide.ml
OCAMLOPT  tactics/autorewrite.ml
OCAMLOPT -a -o tactics/tactics.cmxa
OCAMLOPT  vernac/vernacexpr.ml
OCAMLOPT  vernac/g_vernac.ml
OCAMLOPT  vernac/g_proofs.ml
OCAMLOPT  vernac/vernacprop.ml
OCAMLOPT  vernac/himsg.ml
OCAMLOPT  vernac/explainErr.ml
OCAMLOPT  vernac/egramml.ml
OCAMLOPT  vernac/vernacinterp.ml
OCAMLOPT  vernac/ppvernac.ml
OCAMLOPT  vernac/proof_using.ml
OCAMLOPT  vernac/lemmas.ml
OCAMLOPT  vernac/class.ml
OCAMLOPT  vernac/metasyntax.ml
OCAMLOPT  vernac/auto_ind_decl.ml
OCAMLOPT  vernac/indschemes.ml
OCAMLOPT  vernac/declareDef.ml
OCAMLOPT  vernac/obligations.ml
OCAMLOPT  vernac/comDefinition.ml
OCAMLOPT  vernac/comAssumption.ml
OCAMLOPT  vernac/comInductive.ml
OCAMLOPT  vernac/comFixpoint.ml
OCAMLOPT  vernac/comProgramFixpoint.ml
OCAMLOPT  vernac/classes.ml
OCAMLOPT  vernac/record.ml
OCAMLOPT  vernac/assumptions.ml
OCAMLOPT  vernac/mltop.ml
OCAMLOPT  vernac/vernacentries.ml
OCAMLOPT  vernac/misctypes.ml
OCAMLOPT -a -o vernac/vernac.cmxa
OCAMLOPT  stm/vernac_classifier.ml
OCAMLOPT  stm/stm.ml
OCAMLOPT  stm/proofBlockDelimiter.ml
OCAMLOPT  stm/vio_checking.ml
OCAMLOPT -a -o stm/stm.cmxa
OCAMLOPT  toplevel/vernac.ml
OCAMLOPT  toplevel/coqinit.ml
OCAMLOPT  toplevel/coqargs.ml
OCAMLOPT  toplevel/g_toplevel.ml
OCAMLOPT  toplevel/coqloop.ml
OCAMLOPT  toplevel/coqtop.ml
OCAMLOPT  toplevel/workerLoop.ml
OCAMLOPT -a -o toplevel/toplevel.cmxa
COQMKTOP -o bin/coqtop.opt.exe
COQMKTOP -o bin/coqproofworker.opt.exe
COQMKTOP -o bin/coqtacticworker.opt.exe
COQMKTOP -o bin/coqqueryworker.opt.exe
rm -f bin/coqtop.exe && cp bin/coqtop.opt.exe bin/coqtop.exe
MD5SUM cic.mli
COQMKTOP -o bin/coqidetop.opt.exe
OCAMLOPT  plugins/ltac/tacexpr.ml
OCAMLOPT  plugins/ltac/tacarg.ml
OCAMLOPT  plugins/ltac/tacsubst.ml
OCAMLOPT  plugins/ltac/tacenv.ml
OCAMLOPT  plugins/ltac/pptactic.ml
OCAMLOPT  plugins/ltac/pltac.ml
OCAMLOPT  plugins/ltac/taccoerce.ml
OCAMLOPT  plugins/ltac/tactic_debug.ml
OCAMLOPT  plugins/ltac/tacintern.ml
OCAMLOPT  plugins/ltac/tacentries.ml
OCAMLOPT  plugins/ltac/profile_ltac.ml
OCAMLOPT  plugins/ltac/tactic_matching.ml
OCAMLOPT  plugins/ltac/tacinterp.ml
OCAMLOPT  plugins/ltac/evar_tactics.ml
OCAMLOPT  plugins/ltac/tactic_option.ml
OCAMLOPT  plugins/ltac/extraargs.ml
OCAMLOPT  plugins/ltac/g_obligations.ml
OCAMLOPT  plugins/ltac/coretactics.ml
OCAMLOPT  plugins/ltac/extratactics.ml
OCAMLOPT  plugins/ltac/profile_ltac_tactics.ml
OCAMLOPT  plugins/ltac/g_auto.ml
OCAMLOPT  plugins/ltac/g_class.ml
OCAMLOPT  plugins/ltac/rewrite.ml
OCAMLOPT  plugins/ltac/g_rewrite.ml
OCAMLOPT  plugins/ltac/g_eqdecide.ml
OCAMLOPT  plugins/ltac/g_tactic.ml
OCAMLOPT  plugins/ltac/g_ltac.ml
OCAMLOPT -pack -o plugins/ltac/ltac_plugin.cmx
OCAMLOPT -shared -o plugins/ltac/ltac_plugin.cmxs
OCAMLOPT  plugins/ltac/tauto.ml
OCAMLOPT -pack -o plugins/ltac/tauto_plugin.cmx
OCAMLOPT -shared -o plugins/ltac/tauto_plugin.cmxs
OCAMLOPT  plugins/omega/coq_omega.ml
OCAMLOPT  plugins/omega/g_omega.ml
OCAMLOPT -pack -o plugins/omega/omega_plugin.cmx
OCAMLOPT -shared -o plugins/omega/omega_plugin.cmxs
OCAMLOPT  plugins/romega/refl_omega.ml
OCAMLOPT  plugins/romega/g_romega.ml
OCAMLOPT -pack -o plugins/romega/romega_plugin.cmx
OCAMLOPT -shared -o plugins/romega/romega_plugin.cmxs
OCAMLOPT  plugins/micromega/coq_micromega.ml
OCAMLOPT  plugins/micromega/g_micromega.ml
OCAMLOPT -pack -o plugins/micromega/micromega_plugin.cmx
OCAMLOPT -shared -o plugins/micromega/micromega_plugin.cmxs
OCAMLOPT  plugins/quote/quote.ml
OCAMLOPT  plugins/quote/g_quote.ml
OCAMLOPT -pack -o plugins/quote/quote_plugin.cmx
OCAMLOPT -shared -o plugins/quote/quote_plugin.cmxs
OCAMLOPT  plugins/setoid_ring/newring_ast.ml
OCAMLOPT  plugins/setoid_ring/newring.ml
OCAMLOPT  plugins/setoid_ring/g_newring.ml
OCAMLOPT -pack -o plugins/setoid_ring/newring_plugin.cmx
OCAMLOPT -shared -o plugins/setoid_ring/newring_plugin.cmxs
OCAMLOPT  plugins/extraction/table.ml
OCAMLOPT  plugins/extraction/mlutil.ml
OCAMLOPT  plugins/extraction/modutil.ml
OCAMLOPT  plugins/extraction/extraction.ml
OCAMLOPT  plugins/extraction/common.ml
OCAMLOPT  plugins/extraction/ocaml.ml
OCAMLOPT  plugins/extraction/haskell.ml
OCAMLOPT  plugins/extraction/scheme.ml
OCAMLOPT  plugins/extraction/json.ml
OCAMLOPT  plugins/extraction/extract_env.ml
OCAMLOPT  plugins/extraction/g_extraction.ml
OCAMLOPT -pack -o plugins/extraction/extraction_plugin.cmx
OCAMLOPT -shared -o plugins/extraction/extraction_plugin.cmxs
OCAMLOPT  plugins/cc/cctac.ml
OCAMLOPT  plugins/cc/g_congruence.ml
OCAMLOPT -pack -o plugins/cc/cc_plugin.cmx
OCAMLOPT -shared -o plugins/cc/cc_plugin.cmxs
OCAMLOPT  plugins/firstorder/sequent.ml
OCAMLOPT  plugins/firstorder/rules.ml
OCAMLOPT  plugins/firstorder/instances.ml
OCAMLOPT  plugins/firstorder/ground.ml
OCAMLOPT  plugins/firstorder/g_ground.ml
OCAMLOPT -pack -o plugins/firstorder/ground_plugin.cmx
OCAMLOPT -shared -o plugins/firstorder/ground_plugin.cmxs
OCAMLOPT  plugins/rtauto/refl_tauto.ml
OCAMLOPT  plugins/rtauto/g_rtauto.ml
OCAMLOPT -pack -o plugins/rtauto/rtauto_plugin.cmx
OCAMLOPT -shared -o plugins/rtauto/rtauto_plugin.cmxs
OCAMLOPT  plugins/btauto/refl_btauto.ml
OCAMLOPT  plugins/btauto/g_btauto.ml
OCAMLOPT -pack -o plugins/btauto/btauto_plugin.cmx
OCAMLOPT -shared -o plugins/btauto/btauto_plugin.cmxs
OCAMLOPT  plugins/funind/indfun_common.ml
OCAMLOPT  plugins/funind/glob_termops.ml
OCAMLOPT  plugins/funind/recdef.ml
OCAMLOPT  plugins/funind/glob_term_to_relation.ml
OCAMLOPT  plugins/funind/functional_principles_proofs.ml
OCAMLOPT  plugins/funind/functional_principles_types.ml
OCAMLOPT  plugins/funind/invfun.ml
OCAMLOPT  plugins/funind/indfun.ml
OCAMLOPT  plugins/funind/g_indfun.ml
OCAMLOPT -pack -o plugins/funind/recdef_plugin.cmx
OCAMLOPT -shared -o plugins/funind/recdef_plugin.cmxs
OCAMLOPT  plugins/nsatz/nsatz.ml
OCAMLOPT  plugins/nsatz/g_nsatz.ml
OCAMLOPT -pack -o plugins/nsatz/nsatz_plugin.cmx
OCAMLOPT -shared -o plugins/nsatz/nsatz_plugin.cmxs
OCAMLOPT  plugins/syntax/nat_syntax.ml
OCAMLOPT -pack -o plugins/syntax/nat_syntax_plugin.cmx
OCAMLOPT -shared -o plugins/syntax/nat_syntax_plugin.cmxs
OCAMLOPT  plugins/syntax/positive_syntax.ml
OCAMLOPT -pack -o plugins/syntax/positive_syntax_plugin.cmx
OCAMLOPT -shared -o plugins/syntax/positive_syntax_plugin.cmxs
OCAMLOPT  plugins/syntax/n_syntax.ml
OCAMLOPT -pack -o plugins/syntax/n_syntax_plugin.cmx
OCAMLOPT -shared -o plugins/syntax/n_syntax_plugin.cmxs
OCAMLOPT  plugins/syntax/z_syntax.ml
OCAMLOPT -pack -o plugins/syntax/z_syntax_plugin.cmx
OCAMLOPT -shared -o plugins/syntax/z_syntax_plugin.cmxs
OCAMLOPT  plugins/syntax/r_syntax.ml
OCAMLOPT -pack -o plugins/syntax/r_syntax_plugin.cmx
OCAMLOPT -shared -o plugins/syntax/r_syntax_plugin.cmxs
OCAMLOPT  plugins/syntax/int31_syntax.ml
OCAMLOPT -pack -o plugins/syntax/int31_syntax_plugin.cmx
OCAMLOPT -shared -o plugins/syntax/int31_syntax_plugin.cmxs
OCAMLOPT  plugins/syntax/ascii_syntax.ml
OCAMLOPT -pack -o plugins/syntax/ascii_syntax_plugin.cmx
OCAMLOPT -shared -o plugins/syntax/ascii_syntax_plugin.cmxs
OCAMLOPT  plugins/syntax/string_syntax.ml
OCAMLOPT -pack -o plugins/syntax/string_syntax_plugin.cmx
OCAMLOPT -shared -o plugins/syntax/string_syntax_plugin.cmxs
OCAMLOPT  plugins/derive/g_derive.ml
OCAMLOPT -pack -o plugins/derive/derive_plugin.cmx
OCAMLOPT -shared -o plugins/derive/derive_plugin.cmxs
OCAMLOPT  plugins/ssrmatching/ssrmatching.ml
OCAMLOPT  plugins/ssrmatching/g_ssrmatching.ml
OCAMLOPT -pack -o plugins/ssrmatching/ssrmatching_plugin.cmx
OCAMLOPT -shared -o plugins/ssrmatching/ssrmatching_plugin.cmxs
OCAMLOPT  plugins/ssr/ssrprinters.ml
OCAMLOPT  plugins/ssr/ssrcommon.ml
OCAMLOPT  plugins/ssr/ssrtacticals.ml
OCAMLOPT  plugins/ssr/ssrelim.ml
OCAMLOPT  plugins/ssr/ssrview.ml
OCAMLOPT  plugins/ssr/ssrbwd.ml
OCAMLOPT  plugins/ssr/ssrequality.ml
OCAMLOPT  plugins/ssr/ssripats.ml
OCAMLOPT  plugins/ssr/ssrfwd.ml
OCAMLOPT  plugins/ssr/ssrparser.ml
OCAMLOPT  plugins/ssr/ssrvernac.ml
OCAMLOPT -pack -o plugins/ssr/ssreflect_plugin.cmx
OCAMLOPT -shared -o plugins/ssr/ssreflect_plugin.cmxs
cp bin/coqidetop.opt.exe bin/coqidetop.exe
COQC -noinit theories/Init/Notations.v
COQC -noinit theories/Init/Logic.v
COQC -noinit theories/Init/Datatypes.v
COQC -noinit theories/Init/Logic_Type.v
COQC -noinit theories/Init/Specif.v
COQC -noinit theories/Init/Decimal.v
COQC -noinit theories/Init/Nat.v
COQC -noinit theories/Init/Peano.v
COQC -noinit theories/Init/Wf.v
COQC -noinit theories/Init/Tactics.v
COQC -noinit theories/Init/Tauto.v
COQC -noinit theories/Init/Prelude.v
OCAMLC    printing/proof_diffs.ml
OCAMLC -a -o printing/printing.cma
make[1]: Leaving directory '/cygdrive/c/Users/Jim.DESKTOP-MMPNVGH/Documents/coqSource/coq'
```

**Comment by SkySkimmer** [comment]
It's for optimisation. eg ff it inlined a function from proof_diffs in some other module then the other module must be recompiled, or maybe the change made it a good idea to start inlining some function, etc.

I think the option `-opaque` might allow faster incremental compilation (by disabling cross module optimisation) without being as slow as byte compiled Coq.

Note if you're in the development process and compiling just to check type errors you can make `byte` instead as byte compilation does not have this behaviour.

**Comment by gares** [comment]
It is called "cross module optimization" and is a "feature" of OCaml's native compiler that breaks separate compilation, as you have discovered. Inlining is performed at compilation time, instead of link time. At the time this feature was implemented toolchains like  gcc had no LTO, and the amount of inlining a functional language needs to perform to be efficient is very high. Hence this "solution".

`-opaque` was recently added, before that one would have to "hide" `.cmx` files to the compiler in order to prevent cross module inlining.

Personally, I'd be very happy to have `./configure -profile devel` use `-opaque` (to have faster recompilation) and `./configure -profile release` to turn all optimizations on (at the cost of breaking separate compilation).

**Comment by gares** [comment]
If opaque is used and you don't pass -native it should output correct deps

**Comment by rgrinberg** [comment]
Just a note: dune supports `-opaque` (and appropriately adjusts the rules) for builds in the dev profile out of the box.

**Comment by ejgallego** [comment]
Dune 1.1.1 will build Coq with `-opaque` in the default profile.

**Comment by ejgallego** [comment]
I will close this unless anyone opposes, developers needing fancier features can now use the preliminary Dune setup.

**Comment by jfehrle** [comment]
When will Dune be the standard setup?  Where are the instructions for using Dune?  Can we add that to dev/doc?

**Comment by ejgallego** [comment]
An experimental, dev-only setup has been just merged, see `make -f Makefile.dune` for instructions.

There is some documentation in `dev/doc`, but for sure we may want to improve on it.

Testing on Windows is really appreciated.

**Comment by jfehrle** [comment]
> I will close this unless anyone opposes

Wouldn't the normal procedure be to keep the issue open until Dune is a standard, non-experimental part of Coq and someone has verified that it handles incremental builds as expected?  Though it would be appropriate to mark the issue to so others wait for Dune before crafting their own solution.

**Comment by ejgallego** [comment]
Yeah my point is that this is bug seems to be in a "use Dune or likely wontfix", however you are right in that we should verify it, pending on #8421 .

**Comment by ejgallego** [comment]
#8421 works now so windows users can enjoy this feature.

**Comment by ejgallego** [comment]
Dune's `@check` target provides an ever faster mode; I am closing this bug as developers willing to use this functionality have a solution now, given that Windows + Dune should be reasonably stable now.

My impression is that we don't have the manpower to implement `-opaque` or a "check" mode in the make-based system, thus if I am correct no point in opening a bug against the make-based system that will be tagged as WONTFIX.
